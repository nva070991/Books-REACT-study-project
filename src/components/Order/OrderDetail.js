import { useState, React } from 'react'
import MinMax from '../MinMax/MinMax'
import props from './OrderProps'

function OrderDetail(prop) {
  const {
    book,
    callback,
    Delete,
  } = prop
  const [quantity, setBook] = useState(Number(book.quantity))

  const setQuatinty = (qu, id) => {
    setBook(qu)
    callback(String(qu), id)
  }
  return (
    <div style={{
      boxSizing: 'border-box',
      background: 'white',
      height: '300px',
      width: '300px',
      borderRadius: '10px',
      padding: '14px',
      display: 'flex',
      flexDirection: 'column',
    }}
    >
      <div>
        <button
          style={{
            background: 'red', height: '20px', width: '20px', borderRadius: '10px', border: 'none',
          }}
          type="button"
          className="buttons"
          onClick={() => Delete(book.id)}
        >
          X
        </button>

      </div>
      <div>

        <h3 data-testid="h3" style={{ height: '50px' }}>{book.title}</h3>

      </div>
      <div>
        Price
        {' '}
        {Number(book.price)}
        {' '}
        /
        {' '}
        {Number(book.price) * Number(quantity)}
      </div>
      <div style={{ marginTop: '25px' }}>
        Quanity
        {' '}
        {Number(book.quantity)}
      </div>
      <div>
        <a href={Number(book.id)}>Readmore</a>
      </div>
      <div style={{ marginTop: '25px' }}>
        <MinMax
          max={Number(book.rest)}
          current={Number(book.quantity)}
          onChange={(qu) => setQuatinty(qu, Number(book.id))}
        />
      </div>

    </div>
  )
}

OrderDetail.propTypes = props

export default OrderDetail
