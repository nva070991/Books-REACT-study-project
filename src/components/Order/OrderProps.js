import PropTypes from 'prop-types'

export default {
  book: PropTypes.shape({
    id: PropTypes.string,
    title: PropTypes.string,
    price: PropTypes.string,
    rest: PropTypes.string,
    quantity: PropTypes.string,

  }),
  callback: PropTypes.func.isRequired,
  Delete: PropTypes.func,

}
