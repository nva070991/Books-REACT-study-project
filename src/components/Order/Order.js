/* eslint-disable import/no-extraneous-dependencies */
import { useEffect, useState, React } from 'react'
import axios from 'axios'
import OrderDetail from './OrderDetail'
import Navigation from '../Navigation'
import '../../styles/styles.css'

function Order() {
  const [books, setBooks] = useState([])

  const Delete = (id) => {
    setBooks(
      books.filter((book) => (book.id !== id)),
    )
  }

  const Submit = () => {
    window.location.assign('http://localhost:3000/Order')
  }

  const loadBooks = () => {
    axios
      .get('https://api.jsonbin.io/v3/b/6515255a0574da7622b16f01')
      .then((response) => {
        setBooks(response.data.record)
      })
      .catch((err) => {
        console.log(err.message)
        alert('не смогли загрузить данные из API', err.message)
      })
  }

  useEffect(() => {
    loadBooks()
  }, [])

  const setQuatinty = (quantity, id) => {
    setBooks(
      books.map((book) => (Number(book.id) !== id ? book : { ...book, quantity })),
    )
  }

  function countTotal() {
    let totalQu = 0
    let totalPrice = 0
    for (let i = 0; i < books.length; i += 1) {
      totalQu += Number(books[i].quantity)
      totalPrice += (books[i].price * books[i].quantity)
    }
    return { totalQu, totalPrice }
  }

  useEffect(() => {
    countTotal()
  })

  return (
    <div style={{
      background: 'gray',

    }}
    >
      <Navigation />
      <div style={{
        background: 'gray',
        width: '100vw',
        height: '100vh',
        boxSizing: 'border-box',

      }}
      >
        <div
          className="background DivPadding"
          style={{

            display: 'flex',
            flexWrap: 'wrap',
            gap: '9px',
          }}
        >
          {books[0] ? books.map((book) => (
            <OrderDetail
              key={Number(book.id)}
              book={book}
              callback={setQuatinty}
              Delete={Delete}
            />
          )) : (
            <div style={{

              display: 'flex',
              width: '100vw',
              height: '100vh',
              justifyContent: 'center',
              alignItems: 'center',
              fontSize: '25px',

            }}
            >
              Загрузка
            </div>
          )}

        </div>
        <div className="background">
          {' '}
          Total Quanity:
          {' '}
          <b>{countTotal().totalQu}</b>

        </div>
        <div className="background">
          {' '}
          Total Price:
          {' '}
          <b data-testid="total">{countTotal().totalPrice}</b>

        </div>
        <div className="background DivPadding">
          <button className="Buttons" onClick={Submit} type="button">Send order</button>
        </div>

      </div>

    </div>
  )
}

export default Order
