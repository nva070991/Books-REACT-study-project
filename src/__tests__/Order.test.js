import { render, screen } from '@testing-library/react'
import React from 'react'
import { Routes, Route, BrowserRouter } from 'react-router-dom'
import Order from '../components/Order/Order'
import UserContext from '../contexts/Login'

let login = 1
const SetLogin = (l) => { login = l }
let Password = 1
const SetPassword = (l) => { Password = l }

describe('Order component', () => {
  it('should render Order component', () => {
    const view = render(
      <BrowserRouter>
        <UserContext.Provider value={{
          login, SetLogin, Password, SetPassword,
        }}
        >

          <Routes>
            <Route path="/" element={<Order />} />

          </Routes>
        </UserContext.Provider>
      </BrowserRouter>,
    )
    expect(view).toMatchSnapshot()
  })

  it('should render Total', () => {
    render(
      <BrowserRouter>
        <UserContext.Provider value={{
          login, SetLogin, Password, SetPassword,
        }}
        >

          <Routes>
            <Route path="/" element={<Order />} />

          </Routes>
        </UserContext.Provider>
      </BrowserRouter>,
    )

    expect(screen.getByTestId('total')).toHaveTextContent('0')
  })
})
