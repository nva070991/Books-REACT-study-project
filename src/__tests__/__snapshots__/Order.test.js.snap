// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Order component should render Order component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        style="background: gray;"
      >
        <nav
          class="DivPadding"
          style="display: flex; gap: 8px; width: 100vw;"
        >
          <a
            aria-current="page"
            class="Nav active"
            href="/"
          >
            Home
          </a>
          <a
            class="Nav"
            href="/About"
          >
            About
          </a>
          <a
            class="Nav"
            href="/Order"
          >
            Order
          </a>
          <a
            aria-current="page"
            class="Nav active"
            href="/"
          >
            Log out
          </a>
        </nav>
        <div
          style="background: gray; width: 100vw; height: 100vh; box-sizing: border-box;"
        >
          <div
            class="background DivPadding"
            style="display: flex; flex-wrap: wrap; gap: 9px;"
          >
            <div
              style="display: flex; width: 100vw; height: 100vh; justify-content: center; align-items: center; font-size: 25px;"
            >
              Загрузка
            </div>
          </div>
          <div
            class="background"
          >
             
            Total Quanity:
             
            <b>
              0
            </b>
          </div>
          <div
            class="background"
          >
             
            Total Price:
             
            <b
              data-testid="total"
            >
              0
            </b>
          </div>
          <div
            class="background DivPadding"
          >
            <button
              class="Buttons"
              type="button"
            >
              Send order
            </button>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      style="background: gray;"
    >
      <nav
        class="DivPadding"
        style="display: flex; gap: 8px; width: 100vw;"
      >
        <a
          aria-current="page"
          class="Nav active"
          href="/"
        >
          Home
        </a>
        <a
          class="Nav"
          href="/About"
        >
          About
        </a>
        <a
          class="Nav"
          href="/Order"
        >
          Order
        </a>
        <a
          aria-current="page"
          class="Nav active"
          href="/"
        >
          Log out
        </a>
      </nav>
      <div
        style="background: gray; width: 100vw; height: 100vh; box-sizing: border-box;"
      >
        <div
          class="background DivPadding"
          style="display: flex; flex-wrap: wrap; gap: 9px;"
        >
          <div
            style="display: flex; width: 100vw; height: 100vh; justify-content: center; align-items: center; font-size: 25px;"
          >
            Загрузка
          </div>
        </div>
        <div
          class="background"
        >
           
          Total Quanity:
           
          <b>
            0
          </b>
        </div>
        <div
          class="background"
        >
           
          Total Price:
           
          <b
            data-testid="total"
          >
            0
          </b>
        </div>
        <div
          class="background DivPadding"
        >
          <button
            class="Buttons"
            type="button"
          >
            Send order
          </button>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
